### DISCUSSION TITLE
独自ノーマライズに該当する単語の抽出方法？

### DISCUSSION THREAD

**Discussion by askdkc** [pgroonga/pgroonga#351] [Q&A]
## やりたいこと

https://github.com/pgroonga/pgroonga/discussions/349 の流れからの追加の質問となるのですが、下記のように`pgroonga_highlight_html`でヒットするキーワードがある時に、該当のパターンを返す方法はありますでしょうか？

```sql
select pgroonga_highlight_html(name, ARRAY['斉藤'], 'pgroonga_names_index') from names where name &@~ '斉藤';
--         pgroonga_highlight_html        
-- ---------------------------------------
--  <span class="keyword">斉藤</span>さん
--  <span class="keyword">齊藤</span>さん
--  <span class="keyword">斎藤</span>さん
-- (3 rows)
``` 

具体的には☝️の場合に

```sql
--  <span class="keyword">斉藤</span>さん
--  <span class="keyword">齊藤</span>さん
--  <span class="keyword">斎藤</span>さん
```

ではなく

```sql
{'斉藤','齋藤','齊藤','斎藤'}
```

みたいな全パターンが返って来て欲しい。

恐らく各カラムに付与した独自ノーマライズインデックスに対してではなく`pgroonga_synonyms_index`側に検索キーワードを渡し、該当する全ての漢字の組み合わせが返ってくるような仕組み、もしくはやり方があるかを知りたいです。

## 理由

WEBアプリケーションで検索クエリ実行時に`pgroonga_highlight_html`まで含めて実行すると、その検索クエリがシーケンシャルサーチになってしまい`PGroonga`のインデックスが使われずに遅い（ページネーションする時に都度select count(*)が走るのも原因？）。

それを回避するため（検索時にインデックスを使うため）`pgroonga_highlight_html`を使わずに検索を実行した後、結果一覧に表示させる各項目を検索キーワードの`pgroonga_highlight_html`関数でチェックし、ハイライト表示をさせる形にすると速くなるが、今度はヒットしたデータ件数 x 表示させる項目数に応じて`N+1`でハイライトのためのクエリがDBに発行される（とは言えシーケンシャルサーチよりは当然速いが、DBへのリクエスト数は画面に出す検索結果数の数に応じて膨大になる）。

これで十分速いなら良いのでは？と妥協しても良いのですが、更なる対策として検索キーワードで対象となるノーマライザでヒットする組み合わせ、今回の場合`{'斉藤','齋藤','齊藤','斎藤'} `を一回のみ取得し、それをビュー側に渡すことによって`jQuery`等を使って、検索結果一覧の項目のテキストの中からマッチングする文字列だけにハイライトをかける方が、当然DBへのリクエストも1回で終わり無駄が少ない。（かつN+1みたいなクエリが裏で走ってないので気分が良いかも）

**Comment by kou** [comment]
いやぁ、それはできないっす。
できるようにもしないです。効率よく実現できないんですよ。



> WEBアプリケーションで検索クエリ実行時に`pgroonga_highlight_html`まで含めて実行すると、その検索クエリがシーケンシャルサーチになってしまい`PGroonga`のインデックスが使われずに遅い（ページネーションする時に都度select count(*)が走るのも原因？）。

え、`pgroonga_highlight_html()`を使った出力処理だけがシーケンシャルで、検索部分はPGroongaのインデックスを使いませんか？

> それを回避するため（検索時にインデックスを使うため）`pgroonga_highlight_html`を使わずに検索を実行した後、結果一覧に表示させる各項目を検索キーワードの`pgroonga_highlight_html`関数でチェックし、ハイライト表示をさせる形にすると速くなるが、今度はヒットしたデータ件数 x 表示させる項目数に応じて`N+1`でハイライトのためのクエリがDBに発行される（とは言えシーケンシャルサーチよりは当然速いが、DBへのリクエスト数は画面に出す検索結果数の数に応じて膨大になる）。

```sql
select pgroonga_highlight_html(name, ARRAY['斉藤'], 'pgroonga_names_index')
 from (select name where names &@~ '斉藤' limit 100) as filtered_names;
```

みたいに`limit`してから`limit`分だけ`pgroonga_highlight_html()`するのがいいんですが、ORマッパー経由でページネーションを実装しているとこんなクエリーを発行できないかもしれないんですよねぇ。

**Reply by askdkc** [reply]
> できるようにもしないです。効率よく実現できないんですよ。

凄くわかる気がする。。。

> ORマッパー経由でページネーションを実装しているとこんなクエリーを発行できないかもしれないんですよねぇ。

そしてこれなんですよね。
世の中のウェブアプリケーションはフレームワークのORM使うパターンが基本なので、DB側が早くてもフレームワーク側でボトルネックを生み出す。うむむ。

悩ましい😱
